File: my-folder/0042-trapping-rain-water/solution.java
class Solution {
    public int trap(int[] height) {
        int n = height.length;
        int left = 1;
        int right = n - 2;
        int maxLeft = height[0];
        int maxRight = height[n - 1];
        int water = 0;
        int res = 0;
        while (left <= right) {
            if (maxLeft <= maxRight) {
                water = maxLeft - height[left];
                maxLeft = Math.max(maxLeft, height[left]);
                left++;
            }
            else {
                water = maxRight - height[right];
                maxRight = Math.max(maxRight, height[right]);
                right--;
            }
            if (water > 0) {
                res += water;
            }
        }
        return res;
    }
}

File: my-folder/0042-trapping-rain-water/solution.py
class Solution:
    def trap(self, height: List[int]) -> int:
        '''
        Keep the wall on the 2 sides and flow water into lower parts
        Consider the lower side of the wall with the current index
        If it can hold water, add it to the result 
        Update the wall size
        '''

        res = 0
        left = 1
        right = len(height) - 2
        maxLeft = height[0]
        maxRight = height[-1]
        water = 0
        while left <= right:
            if maxLeft <= maxRight:
                water = maxLeft - height[left]
                maxLeft = max(maxLeft, height[left])
                left += 1
            else:
                water = maxRight - height[right]
                maxRight = max(maxRight, height[right])
                right -= 1
            if water > 0:
                res += water
        
        return res
